delivery_purchase_label/models/purchase_order.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from odoo import SUPERUSER_ID, _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    delivery_label_picking_id = fields.Many2one(
+        "stock.picking",
+        "Delivery Label Picking",
+        store=True,
+        readonly=True,
+    )
+
+    @api.model
+    def _states_to_generate_delivery_label(self):
+        """Labels will be (re)generated only if the PO is in one of these states."""
+        return ["draft", "sent"]
+
+    @api.model
+    def _mail_templates_to_not_attach_labels(self):
+        return self.env.ref("purchase.email_template_edi_purchase_reminder").ids
+
+    def action_rfq_send(self):
+        self.ensure_one()
+        self._generate_purchase_delivery_label()
+        return super().action_rfq_send()
+
+    def _is_valid_for_vendor_labels(self):
+        self.ensure_one()
+        if self.state not in self._states_to_generate_delivery_label():
+            return
+        if not self.dest_address_id:
+            return False
+        if not any(
+            product.type in ["product", "consu"]
+            for product in self.order_line.product_id
+        ):
+            return False
+        return True
+
+    def _is_picking_label_uptodate(self):
+        """Check if the picking label needs to be regenerated."""
+        self.ensure_one()
+        if not self.delivery_label_picking_id:
+            return False
+        pick = self.delivery_label_picking_id
+        moves_new = []
+        for line in self.order_line:
+            for value in line._prepare_stock_moves(self.delivery_label_picking_id):
+                moves_new.append(
+                    {key: value[key] for key in ("product_id", "product_uom_qty")}
+                )
+        move_prev = [
+            {"product_id": move.product_id.id, "product_uom_qty": move.product_uom_qty}
+            for move in pick.move_lines
+        ]
+        if len(move_prev) == len(moves_new):
+            if [
+                move_value for move_value in move_prev if move_value not in moves_new
+            ] == []:
+                return True
+        return False
+
+    def _generate_purchase_delivery_label(self):
+        """Create a transfer to generate the carrier labels."""
+        self.ensure_one()
+        if not self._is_valid_for_vendor_labels():
+            return
+        # Find the carrier that will be used
+        carrier = self.partner_id.purchase_delivery_carrier_id
```

Comment:
We need a carrier on the PO as initially stated. Use cases showed that the vendor will not always ship with the same carrier. You can keep partner `purchase_delivery_carrier_id` as a fallback value but ideally we shouldn't configure any partner. We are generating drop-shipping PO and the PO carrier should come from the SO delivery carrier.
I would define a field on the PO like `vendor_label_carrier_id`.
This new field should be filled in when the PO is created/updated (done outside this module).
```suggestion
        carrier = self.vendor_label_carrier_id or self.partner_id.purchase_delivery_carrier_id
```